### DISCUSSION TITLE
Change USB MIDI device name

### DISCUSSION THREAD

**Discussion by jhsa** [adafruit/Adafruit_TinyUSB_Arduino#199] [Q&A]
Hi, is it possible to change the USB MIDI device name? If so, how?
Thank you in advance for your answers.

**Comment by JonnyHaystack** [comment]
When you say device name, do you just mean the string descriptor? Because you can change that with `USBDevice.setManufacturerDescriptor("<manufacturer name>")` and `USBDevice.setProductDescriptor("<product name>")`, regardless of device class

**Reply by jhsa** [reply]
Also, USBDevice. doesn't compile. it says it wasn't declared

**Reply by JonnyHaystack** [reply]
Hmm, try `TinyUSBDevice` instead? If you don't have USE_TINYUSB defined anywhere it won't have that alias which is fine
![image](https://user-images.githubusercontent.com/1266473/185934198-0c45ab48-44ba-487e-a877-58881f01a914.png)

**Reply by jhsa** [reply]
This is what I have been trying. Where should I place the piece of code you have just posted?

```
#include <Arduino.h>
#include <Adafruit_TinyUSB.h>
#include <MIDI.h>

// USB MIDI object
Adafruit_USBD_MIDI usb_midi;

// Create a new instance of the Arduino MIDI Library,
// and attach usb_midi as the transport.
MIDI_CREATE_INSTANCE(Adafruit_USBD_MIDI, usb_midi, MIDI);


void setup()
{
  
//  TinyUSBDevice.setStringDescriptor("<TinyUSB MIDI>");
//  usb_midi.setStringDescriptor("<TinyUSB MIDI>");
  TinyUSBDevice.setProductDescriptor("TinyUSB MIDI");
//usb_midi.setProductDescriptor("TinyUSB MIDI");


  // Initialize MIDI, and listen to all MIDI channels
  // This will also call usb_midi's begin()

  MIDI.begin(MIDI_CHANNEL_OMNI);

  // Attach the handleNoteOn function to the MIDI Library. It will
  // be called whenever the Bluefruit receives MIDI Note On messages.
  MIDI.setHandleNoteOn(handleNoteOn);

  // Do the same for MIDI Note Off messages.
  MIDI.setHandleNoteOff(handleNoteOff);

  Serial.begin(115200);

  // wait until device mounted
  while( !TinyUSBDevice.mounted() ) delay(1);
}
```

**Comment by jhsa** [comment]
Thanks for your answer. 
I am using an ESP32-S2 Mini board, connect it to the computer's USB port, open the MIDI Monitor software, and see the default MIDI device In/Out ports name for this device is "LOLIN-S2-MINI". This is what I am trying to change. I am using the "midi_test" example sketch.

**Comment by jhsa** [comment]
Definitely not working.. It coimpiles but it doesn't change anything.

```
TinyUSBDevice.setManufacturerDescriptor("<manufacturer name>");
TinyUSBDevice.setProductDescriptor("<product name>");
```

**Reply by JonnyHaystack** [reply]
The OS will always cache the string descriptor for faster device enumeration. If you're on Windows you can uninstall the device through device manager to reset it.

**Reply by jhsa** [reply]
Just did it. it still display the old name :(

**Comment by Petorrrrr** [comment]
I'm experiencing the same problem with a WEMOS S2 Mini. 

Digging around, I found these lines in src/arduino/Adafruit_USBD_Device.cpp that seem to indicate that you can't set this. But this is no longer true, as with some other TinyUSB library found it could be done, so the below might be outdated and I really want to use this library, as it's much better documented.
```
// EPS32 use built-in core descriptor builder.
// Therefore most of descriptors are stubs only
```

Does someone has a solution for this? Or only the next step perhaps?

**Reply by luan007** [reply]
yeah.. as the library uses esp-arduino's core usb.h (located at: Library/Arduino15/packages/esp32/hardware/esp32/2.0.6/cores/esp32/USB.cpp on my mac), you may consider modify the file directly (#define USB_MANUFACTURER and so on).

i.e add these lines on top of that file:

``` C
#define USB_MANUFACTURER "YOUR_NAME"
#define USB_PRODUCT "YOUR_PRODUCT"
```
